In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# ml libraries
from sklearn.svm import SVR
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,StandardScaler
from sklearn.metrics import mean_absolute_error,r2_score,mean_squared_error

In [3]:
df = sns.load_dataset("diamonds")
df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

# 💎 Diamonds Dataset - Variable Description

| Variable | Type        | Description |
|----------|------------|-------------|
| **carat** | Continuous | Weight of the diamond |
| **cut**   | Categorical (Ordinal) | Quality of the cut (Fair, Good, Very Good, Premium, Ideal) |
| **color** | Categorical (Ordinal) | Diamond color, from J (worst) to D (best) |
| **clarity** | Categorical (Ordinal) | Diamond clarity: I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best) |
| **depth** | Continuous | Total depth % = 2 * z / (x + y) |
| **table** | Continuous | Width of the top of the diamond relative to the widest point |
| **price** | Continuous (Target) | Price in US dollars (\$326 – \$18,823) |
| **x** | Continuous | Length in mm |
| **y** | Continuous | Width in mm |
| **z** | Continuous | Depth in mm |


In [4]:
df.shape

(53940, 10)

In [5]:
df.describe()

carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   carat    53940 non-null  float64 
 1   cut      53940 non-null  category
 2   color    53940 non-null  category
 3   clarity  53940 non-null  category
 4   depth    53940 non-null  float64 
 5   table    53940 non-null  float64 
 6   price    53940 non-null  int64   
 7   x        53940 non-null  float64 
 8   y        53940 non-null  float64 
 9   z        53940 non-null  float64 
dtypes: category(3), float64(6), int64(1)
memory usage: 3.0 MB


In [7]:
df['color'].value_counts()

color
G    11292
E     9797
F     9542
H     8304
D     6775
I     5422
J     2808
Name: count, dtype: int64

In [8]:
df['clarity'].value_counts()

clarity
SI1     13065
VS2     12258
SI2      9194
VS1      8171
VVS2     5066
VVS1     3655
IF       1790
I1        741
Name: count, dtype: int64

In [9]:
# Initialize encoder
le = LabelEncoder()

# Loop through columns
for col in df.columns:
    if df[col].dtype == 'object' or df[col].dtype.name == 'category':  # categorical check
        df[col] = le.fit_transform(df[col])


In [10]:
df.head()

carat  cut  color  clarity  depth  table  price     x     y     z
0   0.23    2      1        3   61.5   55.0    326  3.95  3.98  2.43
1   0.21    3      1        2   59.8   61.0    326  3.89  3.84  2.31
2   0.23    1      1        4   56.9   65.0    327  4.05  4.07  2.31
3   0.29    3      5        5   62.4   58.0    334  4.20  4.23  2.63
4   0.31    1      6        3   63.3   58.0    335  4.34  4.35  2.75

In [11]:
X = df.drop("price", axis=1)   # features
y = df["price"]  # target

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# StandardScaler object
scaler = StandardScaler()

# Fit on train data, transform both train and test
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [12]:
# Call the model
model = SVR()

# Train the model
model.fit(X_train_scaled, y_train)

# Test prediction
y_pred = model.predict(X_test_scaled)

In [13]:
# Test prediction
y_pred = model.predict(X_test_scaled)

In [14]:
print(r2_score(y_test,y_pred))

0.5145288727741113
